packages/user/src/models/user.ts

```diff
@@ -3,16 +3,87 @@ import {
   Entity,
   OnInit,
   PrimaryKey,
+  Property,
+  Index,
+  OptionalProps,
+  Filter,
 } from "@mikro-orm/core"
 
-import { generateEntityId } from "@medusajs/utils"
+import { DALUtils, generateEntityId } from "@medusajs/utils"
+import { DAL } from "@medusajs/types"
+import { createPsqlIndexStatementHelper } from "@medusajs/utils"
 
+const userEmailIndexName = "IDX_user_email"
+const userEmailIndexStatement = createPsqlIndexStatementHelper({
+  name: userEmailIndexName,
+  tableName: "user",
+  columns: "email",
+  where: "deleted_at IS NULL",
+})
+
+const userDeletedAtIndexName = "IDX_user_deleted_at"
+const userDeletedAtIndexStatement = createPsqlIndexStatementHelper({
+  name: userDeletedAtIndexName,
+  tableName: "user",
+  columns: "deleted_at",
+})
+
+type OptionalFields =
+  | "first_name"
+  | "last_name"
+  | "metadata"
+  | "avatar_url"
+  | DAL.SoftDeletableEntityDateColumns
 
 @Entity()
+@Filter(DALUtils.mikroOrmSoftDeletableFilterOptions)
 export default class User {
+  [OptionalProps]?: OptionalFields
+
   @PrimaryKey({ columnType: "text" })
   id!: string
 
+  @Property({ columnType: "text", nullable: true })
```

Comment:
todo(maybe): if a user is searchable by text search then add GIN index on first and last name. but you tell me